lib/mongo/topology.ex

```diff
@@ -0,0 +1,216 @@
+            heartbeat_frequency_ms: heartbeat_frequency_ms,
+            opts: opts,
+            monitors: %{},
+            connection_pools: %{}
+          }
+          |> reconcile_servers
+        {:ok, state}
+    end
+  end
+
+  def terminate(_reason, _state) do
+    :ok = Mongo.Events.notify(%TopologyClosedEvent{
+      topology_pid: self
+    })
+  end
+
+  def handle_call(:topology, _from, state) do
+    {:reply, state.topology, state}
+  end
+
+  def handle_call({:connection, address}, _from, state) do
+    {:reply, Map.fetch(state.connection_pools, address), state}
+  end
+
+  # see https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#updating-the-topologydescription
+  def handle_call({:server_description, server_description}, _from, state) do
+    new_state = handle_server_description(state, server_description)
+    if state.topology != new_state.topology do
+      :ok = Mongo.Events.notify(%TopologyDescriptionChangedEvent{
+        topology_pid: self,
+        previous_description: state.topology,
+        new_description: new_state.topology
+      })
+      {:reply, :ok, new_state}
+    else
+      {:reply, :ok, new_state}
+    end
+  end
+
+  def handle_cast({:force_check, server_address}, state) do
+    case Map.fetch(state.monitors, server_address) do
+      {:ok, monitor_pid} ->
+        :ok = Monitor.force_check(monitor_pid)
+        {:noreply, state}
+
+      :error ->
+        # ignore force checks on monitors that don't exist
+        {:noreply, state}
+    end
+  end
+
+  defp handle_server_description(state, server_description) do
+    state
+    |> get_and_update_in([:topology],
+                         &TopologyDescription.update(&1, server_description, length(state.seeds)))
+    |> process_events
+    |> reconcile_servers
+  end
+
+  defp process_events({events, state}) do
+    Enum.each(events, fn
+      {:force_check, _} = message ->
+        :ok = GenServer.cast(self, message)
+      {previous, next} ->
+        if previous != next do
+          :ok = Mongo.Events.notify(%ServerDescriptionChangedEvent{
+            address: next.address,
+            topology_pid: self,
+            previous_description: previous,
+            new_description: next
+          })
+        end
+      _ ->
+        :ok
+    end)
+    state
+  end
+
+  defp reconcile_servers(state) do
+    old_addrs = Map.keys(state.monitors)
+    new_addrs = Map.keys(state.topology.servers)
+    added = new_addrs -- old_addrs
+    removed = old_addrs -- new_addrs
+
+    state = Enum.reduce(added, state, fn (address, state) ->
+      server_description = state.topology.servers[address]
+      connopts = connect_opts_from_address(state.opts, address)
+      heartbeat_frequency = state.heartbeat_frequency_ms
+      args = [
+        server_description,
+        self,
+        heartbeat_frequency,
+        Keyword.put(connopts, :pool, DBConnection.Connection)
```

Comment:
This is overwriting the value of `:pool` that you pass in as configuration. This should probably be `put_new/3`